### ISSUE TITLE
[1.12] Shoggoth Lairs not spawning

### ISSUE THREAD

**Issue by brokenrevolver** [Shinoow/AbyssalCraft#228]
Searched every river and swamp I could find, even changed the generation frequency to 1 from 2. Still can't find any.

**Comment by Shinoow** [comment]
Seems like the generation code doesn't run as frequently as it has in previous MC versions (they should generate, but it seems to be a lot more rare than usual, without the rarity changing at all). I'm gonna conduct some more tests and see what I can do.

**Comment by angelnc1989** [comment]
I can confirm that. Searched like 10000 blocks in all directions, no Shoggoth Lairs.

**Comment by Darkosto** [comment]
I was just about to create an issue for this. Yes, I looked for a while and didn't see them.

**Comment by drakray** [comment]
same in 1.11.2 AbyssalCraft-1.11.2-1.9.4-pre-4, though there is a LOT of structure in overworld Darkland biome

**Comment by janus0** [comment]
Any luck with this? Playing all the mods expert and haven't found any lairs after searcing 10k blocks in various directions.

**Comment by drakray** [comment]
Try changing the config fo4 Lair Spawning to 1, I hot some this way
Also, they are not always easily findable

**Comment by janus0** [comment]
Will do. I just want to say that it's not good having it be extremely rare, though; if i'm not mistaken, this is used to progress out of the beginning of the mod isn't it? I think anything over 500 blocks is excessive.

**Comment by drakray** [comment]
As stated by Shinoow, it's a bug, will be fixed sometime later.
the config is a good workaround.
If you use Biomes o Plenty, some "water" biome can spawn them, but sometime almost entirely underground, so search thoroughly

**Comment by janus0** [comment]
Just to confirm, this works retroactively for a pre-existing world, right?

**Comment by drakray** [comment]
Nop. Unless there is a specific config for that, there is no retrogen on anything

**Comment by Shinoow** [comment]
Putting two and two together, the main reason for the amount of lairs generated plummeting is due to the preconditions checked prior to generation becoming more strict through later updates (the option to limit the chance of generation was originally put in place because of complaints that the lairs were generating too commonly, where I conducted a test in 1.8.9 which had some lairs generating within 100 blocks of each other in areas consisting of multiple connected swamp biomes). The current things limiting the chance of generation (apart from the config option) are countermeasures to ensure that the lairs don't generate above the surface (as in the past they sometimes generated on hills where rivers had run dry between tall biomes) and countermeasures for cascading worldgen lag (which reduces the randomized area in which the structure has a chance of generating).
In order to counter these limitations, I intend on exposing the integer range used for the generation chance, rather than a multiple of 100 (since that's no longer viable due to the aforementioned factors).

**TL;DR**
The config option for the generation chance will be altered to give more power towards changing the rarity of the structure, while also making up for limitations caused by countermeasures put in place to handle factors that negatively affected the generation.

**Comment by Shinoow** [comment]
Whoops, tagged the wrong issues in the commit. This has been resolved as of 1.9.4!
